File: android/src/main/kotlin/com/example/naver_login_flutter/FlutterNaverLoginPlugin.kt
package com.example.naver_login_flutter

import android.app.Activity
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.util.Log
import com.navercorp.nid.NidOAuth
import com.navercorp.nid.oauth.NidOAuthLogin
import com.navercorp.nid.oauth.util.NidOAuthCallback

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ExecutionException

/// 네이버 로그인 상태를 나타내는 열거형
enum class NaverLoginStatus(val value: String) {
    LOGGED_IN("loggedIn"),
    LOGGED_OUT("loggedOut"),
    ERROR("error")
}

/// Flutter 플러그인 메서드를 나타내는 열거형
enum class FlutterPluginMethod {
    InitSdk,
    LogIn,
    LogOut,
    LogOutAndDeleteToken,
    GetCurrentAccount,
    GetCurrentAccessToken,
    RefreshAccessTokenWithRefreshToken,
    IsLoggedIn,
    SetLogEnabled,
    Unknown;

    companion object {
        fun fromMethodName(methodName: String): FlutterPluginMethod {
            return when (methodName) {
                "initSdk" -> InitSdk
                "logIn" -> LogIn
                "logOut" -> LogOut
                "logoutAndDeleteToken" -> LogOutAndDeleteToken
                "getCurrentAccount", "getCurrentAcount" -> GetCurrentAccount // 오타 지원
                "getCurrentAccessToken" -> GetCurrentAccessToken
                "refreshAccessTokenWithRefreshToken" -> RefreshAccessTokenWithRefreshToken
                "isLoggedIn" -> IsLoggedIn
                "setLogEnabled" -> SetLogEnabled
                else -> Unknown
            }
        }
    }
}

/** FlutterNaverLoginPlugin */
class FlutterNaverLoginPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    /// The MethodChannel that will facilitate communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    private val mainScope = CoroutineScope(Dispatchers.Main)

    // Must use this activity instead of context (flutterPluginBinding.applicationContext) to avoid AppCompat issue
    private var activity: Activity? = null

    private lateinit var context: Context

    // pendingResult in login function
    // used to call flutter result in launcher
    private var pendingResult: Result? = null

    // 플러그인/SDK 로그 출력 여부. onAttachedToEngine에서 결정되고
    // setLogEnabled 채널 메서드로 런타임에 변경할 수 있다.
    private var isLogEnabled: Boolean = false

    // MARK: - FlutterPlugin

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "naver_login_flutter")
        channel.setMethodCallHandler(this)

        try {
            context.packageName?.let { packageName ->
                val bundle = context.packageManager?.getApplicationInfo(
                    packageName,
                    PackageManager.GET_META_DATA
                )?.metaData

                applyInitialLogSetting(bundle)

                if (bundle != null) {
                    val clientId = bundle.getString("com.naver.sdk.clientId")
                    val clientSecret = bundle.getString("com.naver.sdk.clientSecret")
                    val clientName = bundle.getString("com.naver.sdk.clientName")

                    logD("=== Naver Login Plugin Registration ===")
                    logD("ClientID: $clientId")
                    logD("ClientSecret: ${clientSecret.masked()}")
                    logD("ClientName: $clientName")
                    logD("===================================")

                    if (clientId != null && clientSecret != null && clientName != null) {
                        try {
                            NidOAuth.initialize(context, clientId, clientSecret, clientName)
                            logD("Naver Login SDK initialized successfully on plugin registration")
                        } catch (e: Exception) {
                            Log.e(LOG_TAG, "Failed to initialize Naver Login SDK: ${e.message}", e)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(LOG_TAG, "Error reading AndroidManifest.xml meta-data: ${e.message}", e)
        }
    }

    /// 로그 초기값을 결정한다.
    /// AndroidManifest의 com.naver.sdk.logEnabled 가 있으면 그 값을,
    /// 없으면 호스트 앱의 debuggable 여부를 따른다.
    private fun applyInitialLogSetting(bundle: android.os.Bundle?) {
        val debuggable = (context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
        setLogEnabled(bundle?.getBoolean("com.naver.sdk.logEnabled", debuggable) ?: debuggable)
    }

    private fun setLogEnabled(enabled: Boolean) {
        isLogEnabled = enabled
        NidOAuth.setLogEnabled(enabled)
    }

    private fun logD(message: String) {
        if (isLogEnabled) Log.d(LOG_TAG, message)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    // MARK: - ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // MARK: - MethodCallHandler

    override fun onMethodCall(call: MethodCall, result: Result) {
        logD("Called method: ${call.method}")

        when (FlutterPluginMethod.fromMethodName(call.method)) {
            FlutterPluginMethod.InitSdk -> {
                @Suppress("UNCHECKED_CAST") val args = call.arguments as Map<String, String?>
                val clientId = args["clientId"] as String
                val clientName = args["clientName"] as String
                val clientSecret = args["clientSecret"] as String
                initSdk(result, clientId, clientName, clientSecret)
            }
            FlutterPluginMethod.LogIn -> login(result)
            FlutterPluginMethod.LogOut -> logout(result)
            FlutterPluginMethod.LogOutAndDeleteToken -> logoutAndDeleteToken(result)
            FlutterPluginMethod.GetCurrentAccessToken -> getCurrentAccessToken(result)
            FlutterPluginMethod.GetCurrentAccount -> {
                mainScope.launch {
                    getCurrentAccount(result)
                }
            }
            FlutterPluginMethod.RefreshAccessTokenWithRefreshToken -> refreshAccessTokenWithRefreshToken(result)
            FlutterPluginMethod.IsLoggedIn -> isLoggedIn(result)
            FlutterPluginMethod.SetLogEnabled -> {
                val enabled = (call.arguments as? Map<*, *>)?.get("enabled") as? Boolean ?: false
                setLogEnabled(enabled)
                result.success(null)
            }
            FlutterPluginMethod.Unknown -> result.notImplemented()
        }
    }

    // MARK: - Private Methods

    private fun initSdk(result: Result, clientId: String, clientName: String, clientSecret: String) {
        try {
            logD("Init SDK")
            logD("- clientId: $clientId")
            logD("- clientName: $clientName")
            logD("- clientSecret: ${clientSecret.masked()}")

            NidOAuth.initialize(context, clientId, clientSecret, clientName)
            sendResult(NaverLoginStatus.LOGGED_OUT, null, null, result)

        } catch (e: Exception) {
            e.printStackTrace()
            result.error(
                e.javaClass.simpleName,
                "NaverIdLoginSDK.initialize failed. message: " + e.localizedMessage,
                null
            )
        }
    }

    private suspend fun getCurrentAccount(result: Result, includeToken: Boolean = false) {
        // SDK 초기화 상태 확인
        try {
            val state = NidOAuth.getState()
        } catch (e: Exception) {
            sendError("SDK not initialized. Please call initSdk first.", result)
            return
        }

        val accessToken = NidOAuth.getAccessToken()

        if (accessToken == null) {
            sendResult(NaverLoginStatus.LOGGED_OUT, null, null, result)
            return
        }

        // 로그인 결과에는 iOS와 동일하게 토큰 정보를 함께 반환한다 (#14)
        val tokenInfo = if (includeToken) mapOf(
            "accessToken" to accessToken,
            "refreshToken" to (NidOAuth.getRefreshToken() ?: ""),
            "tokenType" to (NidOAuth.getTokenType() ?: "bearer"),
            "expiresAt" to formatExpiresAt(NidOAuth.getExpiresAt())
        ) else null

        try {
            val res = getUserInfo(accessToken)
            val obj = JSONObject(res)
            val account = jsonObjectToMap(obj.getJSONObject("response"))
            sendResult(NaverLoginStatus.LOGGED_IN, tokenInfo, account, result)
        } catch (e: InterruptedException) {
            e.printStackTrace()
            sendError("Failed to get user info: ${e.message}", result)
        } catch (e: ExecutionException) {
            e.printStackTrace()
            sendError("Failed to get user info: ${e.message}", result)
        } catch (e: JSONException) {
            e.printStackTrace()
            sendError("Failed to parse user info: ${e.message}", result)
        }
    }



    private fun login(result: Result) {
        // SDK 초기화 상태 확인
        try {
            val state = NidOAuth.getState()
            logD("Current SDK state: $state")
        } catch (e: Exception) {
            sendError("SDK not initialized. Please call initSdk first.", result)
            return
        }

        pendingResult = result

        val mOAuthLoginHandler = object : NidOAuthCallback {
            override fun onSuccess() {
                mainScope.launch {
                    getCurrentAccount(result, includeToken = true)
                }
            }

            override fun onFailure(errorCode: String, errorDesc: String) {
                // 사용자 취소인지 확인
                if (errorCode == "user_cancel" || errorDesc.contains("cancel", ignoreCase = true)) {
                    sendResult(NaverLoginStatus.LOGGED_OUT, null, null, result)
                } else {
                    result.success(object : HashMap<String, String>() {
                        init {
                            put("status", "error")
                            put("errorMessage", "errorCode:$errorCode, errorDesc:$errorDesc")
                        }
                    })
                }
                // Already handled result. We don't need this at the ActivityResult as pending status
                pendingResult = null
            }
        }

        val performLogin = {
            activity?.let {
                NidOAuth.requestLogin(it, mOAuthLoginHandler)
            } ?: run {
                sendError("Activity is null", result)
                pendingResult = null
            }
        }

        // 기존 토큰이 있다면 먼저 삭제 (user_cancel 문제 방지)
        try {
            if (NidOAuth.getAccessToken() != null) {
                logD("Existing token found, logging out first")
                NidOAuth.logout(object : NidOAuthCallback {
                    override fun onSuccess() { performLogin() }
                    override fun onFailure(errorCode: String, errorDesc: String) { performLogin() }
                })
            } else {
                performLogin()
            }
        } catch (e: Exception) {
            // 토큰 체크 실패는 무시하고 계속 진행
            logD("Token check failed: ${e.message}")
            performLogin()
        }
    }

    private fun logout(result: Result) {
        try {
            NidOAuth.logout(object : NidOAuthCallback {
                override fun onSuccess() {
                    sendResult(NaverLoginStatus.LOGGED_OUT, null, null, result)
                }
                override fun onFailure(errorCode: String, errorDesc: String) {
                    sendResult(NaverLoginStatus.LOGGED_OUT, null, null, result)
                }
            })
        } catch (e: Exception) {
            /**
            Firebase Crasylytics error workaround

            ArrayDecoders.decodeUnknownField
            com.google.crypto.tink.shaded.protobuf.c0 - Protocol message contained an invalid tag (zero).
             */
            e.printStackTrace()
            sendResult(NaverLoginStatus.LOGGED_OUT, null, null, result)
        }
    }

    private fun logoutAndDeleteToken(result: Result) {
        val mOAuthLoginHandler = object : NidOAuthCallback {
            override fun onSuccess() {
                sendResult(NaverLoginStatus.LOGGED_OUT, null, null, result)
            }

            override fun onFailure(errorCode: String, errorDesc: String) {
                // 서버에서 token 삭제에 실패했어도 클라이언트에 있는 token 은 삭제되어 로그아웃된 상태이다
                // 실패했어도 클라이언트 상에 token 정보가 없기 때문에 추가적으로 해줄 수 있는 것은 없음
                result.success(object : HashMap<String, String>() {
                    init {
                        put("status", "error")
                        put("errorMessage", "errorCode:$errorCode, errorDesc:$errorDesc")
                    }
                })
            }
        }

        NidOAuth.disconnect(mOAuthLoginHandler)
    }

    private fun getCurrentAccessToken(result: Result) {
        logD("handleGetCurrentAccessToken")

        // SDK 초기화 상태 확인
        try {
            val state = NidOAuth.getState()
        } catch (e: Exception) {
            sendError("SDK not initialized. Please call initSdk first.", result)
            return
        }

        val accessToken = NidOAuth.getAccessToken()
        val refreshToken = NidOAuth.getRefreshToken()
        val expiresAt = NidOAuth.getExpiresAt()
        val tokenType = NidOAuth.getTokenType()

        if (accessToken == null) {
            sendResult(NaverLoginStatus.LOGGED_OUT, null, null, result)
            return
        }

        val tokenInfo = mapOf(
            "accessToken" to accessToken,
            "refreshToken" to (refreshToken ?: ""),
            "tokenType" to (tokenType ?: "bearer"),
            "expiresAt" to formatExpiresAt(expiresAt)
        )

        sendResult(NaverLoginStatus.LOGGED_IN, tokenInfo, null, result)
    }

    private fun refreshAccessTokenWithRefreshToken(result: Result) {
        val refreshToken = NidOAuth.getRefreshToken()
        if (refreshToken == null) {
            sendError("No refresh token available", result)
            return
        }

        val mOAuthLoginHandler = object : NidOAuthCallback {
            override fun onSuccess() {
                val accessToken = NidOAuth.getAccessToken()
                val newRefreshToken = NidOAuth.getRefreshToken()
                val expiresAt = NidOAuth.getExpiresAt()
                val tokenType = NidOAuth.getTokenType()

                if (accessToken != null) {
                    val tokenInfo = mapOf(
                        "accessToken" to accessToken,
                        "refreshToken" to (newRefreshToken ?: ""),
                        "tokenType" to (tokenType ?: "bearer"),
                        "expiresAt" to formatExpiresAt(expiresAt)
                    )
                    sendResult(NaverLoginStatus.LOGGED_IN, tokenInfo, null, result)
                } else {
                    sendError("Failed to get refreshed access token", result)
                }
            }

            override fun onFailure(errorCode: String, errorDesc: String) {
                result.success(object : HashMap<String, String>() {
                    init {
                        put("status", "error")
                        put("errorMessage", "errorCode:$errorCode, errorDesc:$errorDesc")
                    }
                })
            }
        }

        NidOAuthLogin().callRefreshAccessTokenApi(mOAuthLoginHandler)
    }

    private fun isLoggedIn(result: Result) {
        // SDK 초기화 상태 확인
        try {
            val state = NidOAuth.getState()
        } catch (e: Exception) {
            sendError("SDK not initialized. Please call initSdk first.", result)
            return
        }

        val accessToken = NidOAuth.getAccessToken()
        if (accessToken != null) {
            sendResult(NaverLoginStatus.LOGGED_IN, null, null, result)
        } else {
            sendResult(NaverLoginStatus.LOGGED_OUT, null, null, result)
        }
    }

    // MARK: - Helper Methods

    private suspend fun getUserInfo(token: String): String = withContext(Dispatchers.IO) {
        val header = "Bearer $token"
        try {
            val apiURL = "https://openapi.naver.com/v1/nid/me"
            val url = URL(apiURL)
            val con = url.openConnection() as HttpURLConnection
            con.requestMethod = "GET"
            con.setRequestProperty("Authorization", header)
            val responseCode = con.responseCode
            val br: BufferedReader = if (responseCode == 200) {
                BufferedReader(InputStreamReader(con.inputStream))
            } else {
                BufferedReader(InputStreamReader(con.errorStream))
            }
            val response = br.use(BufferedReader::readText)
            br.close()
            response
        } catch (e: Exception) {
            e.printStackTrace()
            ""
        }
    }

    private fun formatExpiresAt(expiresAt: Long): String {
        val date = Date(expiresAt * 1000) // 초를 밀리초로 변환
        val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.getDefault())
        formatter.timeZone = TimeZone.getTimeZone("UTC")
        return formatter.format(date)
    }

    @Throws(JSONException::class)
    fun jsonObjectToMap(jObject: JSONObject): HashMap<String, String> {
        val map = HashMap<String, String>()
        val keys = jObject.keys()

        while (keys.hasNext()) {
            val key = keys.next() as String
            val value = jObject.getString(key)
            map[key] = value
        }
        return map
    }

    // MARK: - Result Handling

    private fun sendResult(
        status: NaverLoginStatus,
        accessToken: Map<String, Any>? = null,
        account: Map<String, String>? = null,
        result: Result
    ) {
        val resultMap = mutableMapOf<String, Any>("status" to status.value.lowercase())

        accessToken?.let { resultMap["accessToken"] = it }
        account?.let { resultMap["account"] = it }

        result.success(resultMap)
    }

    private fun sendError(message: String, result: Result) {
        val errorInfo = mapOf(
            "status" to NaverLoginStatus.ERROR.value.lowercase(),
            "errorMessage" to message
        )

        result.success(errorInfo)
    }
}
private const val LOG_TAG = "NaverLoginFlutter"

/// 값이 들어왔는지만 확인할 수 있게 앞 3자와 길이만 남긴다.
private fun String?.masked(): String = when {
    this.isNullOrEmpty() -> "(not set)"
    length <= 4 -> "*".repeat(length) + " ($length chars)"
    else -> take(3) + "*".repeat(length - 3) + " ($length chars)"
}

File: ios/naver_login_flutter/Sources/naver_login_flutter/FlutterNaverLoginPlugin.swift
import Flutter
import NidThirdPartyLogin
import NidCore
import SafariServices
import UIKit

/// 네이버 로그인 상태를 나타내는 열거형
public enum NaverLoginStatus: String {
    case loggedIn = "loggedIn"
    case loggedOut = "loggedOut"
    case error = "error"
}

/// Flutter 플러그인 메서드를 나타내는 열거형
private enum NaverLoginPluginMethod {
    case initSdk
    case logIn
    case logOut
    case logoutAndDeleteToken
    case getCurrentAccount
    case getCurrentAccessToken
    case refreshAccessTokenWithRefreshToken
    case isLoggedIn
    case unknown
    
    init(methodName: String) {
        switch methodName {
        case "initSdk":
            self = .initSdk
        case "logIn":
            self = .logIn
        case "logOut":
            self = .logOut
        case "logoutAndDeleteToken":
            self = .logoutAndDeleteToken
        case "getCurrentAccount":
            self = .getCurrentAccount
        case "getCurrentAccessToken":
            self = .getCurrentAccessToken
        case "refreshAccessTokenWithRefreshToken":
            self = .refreshAccessTokenWithRefreshToken
        case "isLoggedIn":
            self = .isLoggedIn
        default:
            self = .unknown
        }
    }
}

/// 네이버 로그인 플러그인의 메인 클래스
@objc
public class FlutterNaverLoginPlugin: NSObject, FlutterPlugin, FlutterSceneLifeCycleDelegate {
    private var pendingResult: FlutterResult?

    // MARK: - Lifecycle

    override public init() {
        super.init()
    }

    deinit {
        // Cleanup if needed
    }

    // MARK: - Flutter Plugin Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FlutterNaverLoginPlugin()
        
        // Info.plist에서 네이버 로그인 설정 값 읽기
        let infoDictionary = Bundle.main.infoDictionary
        
        guard let clientId = infoDictionary?["NidClientID"] as? String,
              let clientSecret = infoDictionary?["NidClientSecret"] as? String,
              let appName = infoDictionary?["NidAppName"] as? String,
              let urlScheme = infoDictionary?["NidUrlScheme"] as? String else {
            print("Error: Required Naver Login configuration not found in Info.plist")
            return
        }
        
        // SDK 초기화
        NidOAuth.shared.initialize(
            appName: appName,
            clientId: clientId,
            clientSecret: clientSecret,
            urlScheme: urlScheme
        )
        
        // 기본 로그인 동작 설정 (네이버 앱이 설치된 경우 네이버 앱으로 인증, 네이버 앱이 설치되어있지 않은 경우 SafariViewController를 실행해 인증하는 방식)
        NidOAuth.shared.setLoginBehavior(.appPreferredWithInAppBrowserFallback)
        
        let channel = FlutterMethodChannel(
            name: "naver_login_flutter",
            binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
        // 레거시(AppDelegate) 생명주기: application(_:open:options:) 콜백 수신
        registrar.addApplicationDelegate(instance)
        // UIScene 생명주기(Flutter 신규 템플릿의 FlutterSceneDelegate):
        // scene(_:openURLContexts:) 콜백 수신. 이 경로가 없으면 SceneDelegate 기반
        // 앱에서는 로그인 콜백 URL이 플러그인까지 전달되지 않아 인증이 완료되지 않는다.
        registrar.addSceneDelegate(instance)
    }

    // MARK: - Handle Method Calls

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if pendingResult != nil {
            sendError(message: "Another request is in progress. Please wait", result: result)
            return
        }
        self.pendingResult = result
        let flutterMethod = NaverLoginPluginMethod(methodName: call.method)
        switch flutterMethod {
        case .initSdk:
            guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", 
                              message: "Arguments are required", 
                              details: nil))
            return
            }
            handleInitSdk(args)
        case .logIn:
            handleLogin()
        case .logOut:
            handleLogout()
        case .logoutAndDeleteToken:
            handleLogoutAndDeleteToken()
        case .getCurrentAccount:
            handleGetCurrentAccount()
        case .getCurrentAccessToken:
            handleGetCurrentAccessToken()
        case .refreshAccessTokenWithRefreshToken:
            handleRefreshToken()
        case .isLoggedIn:
            handleIsLoggedIn()
        case .unknown:
            pendingResult?(FlutterMethodNotImplemented)
            pendingResult = nil
        }
    }

    // MARK: - URL Callback Handling

    /// 들어온 URL이 네이버 전용 URL Scheme일 때만 SDK에 전달한다.
    /// - Returns: 네이버가 처리한 경우 `true`. 그 외에는 `false`를 반환해
    ///            다른 딥링크 플러그인이 같은 URL을 이어서 처리할 수 있도록 한다.
    private func handleNaverURL(_ url: URL) -> Bool {
        // 충돌 방지: Info.plist에 등록된 네이버 전용 URL Scheme인지 먼저 확인
        guard let scheme = Bundle.main.infoDictionary?["NidUrlScheme"] as? String,
              url.scheme?.lowercased() == scheme.lowercased() else {
            return false
        }
        return NidOAuth.shared.handleURL(url)
    }

    /// 레거시(UIApplicationDelegate) 생명주기 경로 - AppDelegate 기반 앱
    public func application(_ application: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
        return handleNaverURL(url)
    }

    /// UIScene 생명주기 경로 - SceneDelegate(FlutterSceneDelegate) 기반 앱.
    /// 앱이 실행 중일 때 들어오는 URL은 이 콜백으로만 전달되므로 반드시 처리해야 한다.
    @available(iOS 13.0, *)
    public func scene(_ scene: UIScene, openURLContexts URLContexts: Set<UIOpenURLContext>) -> Bool {
        for context in URLContexts {
            if handleNaverURL(context.url) {
                return true
            }
        }
        return false
    }

    // MARK: - Handler Methods

    private func handleInitSdk(_ args: [String: Any]) {
        // Info.plist에서 네이버 로그인 설정 값 읽기
        let infoDictionary = Bundle.main.infoDictionary
        guard let clientId = infoDictionary?["NidClientID"] as? String,
              let clientSecret = infoDictionary?["NidClientSecret"] as? String,
              let appName = infoDictionary?["NidAppName"] as? String,
              let urlScheme = infoDictionary?["NidUrlScheme"] as? String else {
            sendError(message: "Required Naver Login configuration not found in Info.plist. Please check NidClientID, NidClientSecret, NidAppName, and NidUrlScheme values.")
            return
        }

        // 로그인 동작 설정
        if let behavior = args["loginBehavior"] as? String {
            switch behavior {
            case "inAppBrowser":
                NidOAuth.shared.setLoginBehavior(.inAppBrowser)
            case "app":
                NidOAuth.shared.setLoginBehavior(.app)
            case "appPreferredWithInAppBrowserFallback":
                NidOAuth.shared.setLoginBehavior(.appPreferredWithInAppBrowserFallback)
            default:
                print("Unknown login behavior: \(behavior)")
                break
            }
        }

        // SDK 초기화 완료 후 결과 전송
        sendResult(status: .loggedOut)
    }

    private func handleLogin() {
        NidOAuth.shared.requestLogin { [weak self] result in
            switch result {
            case .success(let loginResult):
                let tokenInfo: [String: Any] = [
                    "accessToken": loginResult.accessToken.tokenString,
                    "refreshToken": loginResult.refreshToken.tokenString,
                    "tokenType": "bearer",
                    "expiresAt": loginResult.accessToken.expiresAt.iso8601String()
                ]
                // 프로필 정보 조회
                self?.getUserProfile(accessToken: loginResult.accessToken.tokenString) { profileResult in
                    switch profileResult {
                    case .success(let profile):
                        self?.sendResult(status: .loggedIn, accessToken: tokenInfo, account: profile)
                    case .failure(let error):
                        self?.sendError(message: error.localizedDescription)
                    }
                }
            case .failure(let error):
                // 에러 메시지를 더 자세히 확인
                if error.localizedDescription.contains("cancel") {
                    self?.sendError(message: "Login cancelled by user")
                } else {
                    self?.sendError(message: error.localizedDescription)
                }
            }
        }
    }

    private func handleLogout() {
        NidOAuth.shared.logout()
        sendResult(status: .loggedOut)
    }

    private func handleLogoutAndDeleteToken() {
        // 먼저 토큰 삭제 시도
        NidOAuth.shared.disconnect { [weak self] result in
            switch result {
            case .success:
                // 토큰 삭제 성공 후 로그아웃 실행
                NidOAuth.shared.logout()
                self?.sendResult(status: .loggedOut)
            case .failure(let error):
                // 토큰 삭제 실패 시에도 로그아웃은 시도
                NidOAuth.shared.logout()
                self?.sendError(message: "Token deletion failed: \(error.localizedDescription)")
            }
        }
    }

    private func handleGetCurrentAccount() {
        guard let accessToken = NidOAuth.shared.accessToken?.tokenString else {
            sendError(message: "No access token available")
            return
        }

        NidOAuth.shared.getUserProfile(accessToken: accessToken) { [weak self] result in
            switch result {
            case .success(let profile):
                self?.sendResult(status: .loggedIn, account: profile)
            case .failure(let error):
                self?.sendError(message: error.localizedDescription)
            }
        }
    }

    private func handleGetCurrentAccessToken() {
        guard let token = NidOAuth.shared.accessToken else {
            sendResult(status: .loggedOut)
            return
        }

        let tokenInfo: [String: Any] = [
            "accessToken": token.tokenString,
            "refreshToken": NidOAuth.shared.refreshToken?.tokenString ?? "",
            "tokenType": "bearer",
            "expiresAt": token.expiresAt.iso8601String()
        ]

        sendResult(status: .loggedIn, accessToken: tokenInfo)
    }

    private func handleRefreshToken() {
        guard NidOAuth.shared.refreshToken != nil else {
            sendError(message: "No refresh token available")
            return
        }

        // SDK는 별도의 silent refresh API를 제공하지 않습니다.
        // reauthenticate는 재인증(UI 표시 가능) 흐름이므로,
        // Flutter 레이어의 refreshAccessTokenWithRefreshToken 메서드와 동일하게 동작합니다.
        NidOAuth.shared.reauthenticate { [weak self] result in
            switch result {
            case .success(let loginResult):
                let tokenInfo: [String: Any] = [
                    "accessToken": loginResult.accessToken.tokenString,
                    "refreshToken": loginResult.refreshToken.tokenString,
                    "tokenType": "bearer",
                    "expiresAt": loginResult.accessToken.expiresAt.iso8601String()
                ]
                self?.sendResult(status: .loggedIn, accessToken: tokenInfo)
            case .failure(let error):
                self?.sendError(message: error.localizedDescription)
            }
        }
    }

    private func handleIsLoggedIn() {
        if NidOAuth.shared.accessToken?.tokenString != nil {
            sendResult(status: .loggedIn)
        } else {
            sendResult(status: .loggedOut)
        }
    }

    // MARK: - Helper Methods

    private func getUserProfile(accessToken: String, completion: @escaping (Result<[String: String], Error>) -> Void) {
        NidOAuth.shared.getUserProfile(accessToken: accessToken) { result in
            completion(result.mapError { $0 as Error })
        }
    }

    // MARK: - Result Handling

    private func sendResult(status: NaverLoginStatus, accessToken: [String: Any]? = nil, account: [String: String]? = nil) {
        var result: [String: Any] = ["status": status.rawValue.lowercased()]
        
        if let accessToken = accessToken {
            result["accessToken"] = accessToken
        }
        
        if let account = account {
            result["account"] = account
        }
        
        DispatchQueue.main.async {
            self.pendingResult?(result)
            self.pendingResult = nil
        }
    }

    private func sendError(message: String, result: FlutterResult? = nil) {
        let errorInfo: [String: Any] = [
            "status": NaverLoginStatus.error.rawValue.lowercased(),
            "errorMessage": message
        ]
        
        DispatchQueue.main.async {
            (result ?? self.pendingResult)?(errorInfo)
            self.pendingResult = nil
        }
    }
}

extension Date {
    func iso8601String() -> String {
        let formatter = ISO8601DateFormatter()
        return formatter.string(from: self)
    }
}
